### PR TITLE
feat(product, types): added missing product variant methods

### PR DESCRIPTION
**Pull request by riqwan**
what:

- adds retrieve and list product variant methods

RESOLVES CORE-1425

### CODE REVIEW

File: packages/product/integration-tests/__tests__/services/product-module-service/product-variants.spec.ts
Comment (adrien2p) [inline packages/product/integration-tests/__tests__/services/product-module-service/product-variants.spec.ts:8]: todo: I suggest to keep that in the module.spec.ts test in a separate describe eventually. what do you think? there is already some tests on the product

File: packages/product/integration-tests/__tests__/services/product-module-service/product-variants.spec.ts
Comment (adrien2p) [inline packages/product/integration-tests/__tests__/services/product-module-service/product-variants.spec.ts]: todo: `should return the variants and count` I don't see the properties, scoped out of the result. Can I get you to apply the same approach to all the tests description as we previously discussed. I will enforce more convention to the tests as it becomes quite difficult to retrieve and read the description which does not really describe the test purpose 🚀

File: packages/product/integration-tests/__tests__/services/product-module-service/product-variants.spec.ts
Comment (adrien2p) [inline packages/product/integration-tests/__tests__/services/product-module-service/product-variants.spec.ts]: question: Should we investigate and include it in this pr?

File: packages/product/integration-tests/__tests__/services/product-module-service/product-variants.spec.ts
Comment (adrien2p) [inline packages/product/integration-tests/__tests__/services/product-module-service/product-variants.spec.ts]: suggestion: `should return the requested variant` Same approach. Some description are missing some words to make it readable as a proper sentence :)

File: packages/product/integration-tests/utils/services/setup-module-service.ts
Comment (adrien2p) [inline packages/product/integration-tests/utils/services/setup-module-service.ts]: todo: this should not be necessary, everything is already setup in the module.spec.ts

File: packages/product/src/services/product-variant.ts
Comment (adrien2p) [inline packages/product/src/services/product-variant.ts]: suggestion: You can use ProductVariant here I think since TEntity must extends at least ProductVariant. wdyt? I am fine keeping it as it is but I wanted to get your feedback.

**Review by adrien2p — CHANGES_REQUESTED** [review]
Very nice, few todo/s and suggestion. I would like us to get better to describe our tests. at the moment we are not consistent and it is what helps someone to understand what we are doing by just reading the tests. I think it is important to start enforcing some consistency, wdyt?

File: packages/product/integration-tests/__tests__/services/product-module-service/product-variants.spec.ts
Comment (riqwan) [reply packages/product/integration-tests/__tests__/services/product-module-service/product-variants.spec.ts:8]: That file will end up being super massive, I'd prefer to keep each separate.

File: packages/product/integration-tests/__tests__/services/product-module-service/product-variants.spec.ts
Comment (riqwan) [reply packages/product/integration-tests/__tests__/services/product-module-service/product-variants.spec.ts]: Not yet, this will be a general issue with pagination. I'll tackle this separately.

File: packages/product/src/services/product-variant.ts
Comment (riqwan) [reply packages/product/src/services/product-variant.ts]: Based on the convo we had yesterday, it would be more aligned to do TEntity here, right?

File: packages/product/integration-tests/__tests__/services/product-module-service/product-variants.spec.ts
Comment (adrien2p) [reply packages/product/integration-tests/__tests__/services/product-module-service/product-variants.spec.ts:8]: I see, I am fine with that I can clean what I ve done in the module tests. We can keep it that way and I ll clean up after at the same time I ll move the tests I have already done

Comment (riqwan) [reply packages/product/integration-tests/__tests__/services/product-module-service/product-variants.spec.ts:8]: No, don't worry about it, i'll end up there anyway, will cleanup once I get to that point.

File: packages/product/integration-tests/__tests__/services/product-module-service/product-variants.spec.ts
Comment (olivermrbl) [inline packages/product/integration-tests/__tests__/services/product-module-service/product-variants.spec.ts]: question: When will we attend to this todo and the skipping of the test?

File: packages/product/src/services/product-variant.ts
Comment (olivermrbl) [reply packages/product/src/services/product-variant.ts]: Curious about this convo. Why are we not typing this entire method with `ProductVariant` instead of TEntity?

Comment (olivermrbl) [reply packages/product/src/services/product-variant.ts]: And other methods

File: packages/product/src/services/product-variant.ts
Comment (olivermrbl) [inline packages/product/src/services/product-variant.ts:28]: comment: In the core, we throw on undefined entity ids in `retrieve`-methods. Do we want to add that either here or in the "gateway" service? I am fine to leave it out, as I think it's reasonable to assume data is passed correctly, but maybe it would make for a better DX.

File: packages/product/integration-tests/__tests__/services/product-module-service/product-variants.spec.ts
Comment (riqwan) [reply packages/product/integration-tests/__tests__/services/product-module-service/product-variants.spec.ts]: I've already begun investigating this, this should be in the next PR.

Comment (riqwan) [reply packages/product/integration-tests/__tests__/services/product-module-service/product-variants.spec.ts]: Didn't add it here since it's a general issue with pagination across services.

File: packages/product/src/services/product-variant.ts
Comment (riqwan) [reply packages/product/src/services/product-variant.ts]: This is something I'm still wrapping my head around.

But the discussion stemmed from the model and repository layer being completely replaceable for which the service should not be tied to model and repository types. There is a public interface that it must obey for repositories `DAL.RepositoryService` for which the service will interface with, but we don't have something similar for models.

If we are passing ProductVariant everywhere, then I don't understand the role of TEntity here and why we can't just use ProductVariant in each case as we are monkey patching them anyway.

Comment (adrien2p) [reply packages/product/src/services/product-variant.ts]: @riqwan the explanation is the exact same as for the repository, which we have discussed again last day 😊

TEntity allows the users to specify which type it will be based on for the input/output, not for us. In short, TEntity extends ProdfuctVariant, therefore TEntity at least contains the properties of product variant but can be instanciated with another type than product variant. If we didn't allow the user to customise the DAL, it wouldn't be necessary, but the fact is that internally we know that tentity is at least product variant, but if the users customise it, then it will be able to return more than product variant, it is for that reason that we can use the entity as the type here while returning TEntity for the user that consume it because of that constraints that the type should extends ProductVariant. 

If it is not clear, we can re discuss that tomorrow wdyt?

Comment (adrien2p) [reply packages/product/src/services/product-variant.ts]: Long story short (summary): we know that TEntity at least extends ProductVariant, we are based on ProductVariant, if a user extends ProductVariant, he can pass it as TEntity which will be the type on whoch the input/output will be based on. Without TEntity, the user will never be able to specify that it is in fact more than ProductVariant and will have to go through a definition file to extend the original type like we explain to do in the core. Don't know if it is shorter 🤣

Comment (riqwan) [reply packages/product/src/services/product-variant.ts]: So, given that explanation, which one would you say we should type `buildQuery` with?

File: packages/product/integration-tests/__tests__/services/product-module-service/product-variants.spec.ts
Comment (olivermrbl) [reply packages/product/integration-tests/__tests__/services/product-module-service/product-variants.spec.ts]: Makes sense! 👍

**Review by olivermrbl — APPROVED** [review]
LGTM, pending final approval from @adrien2p :)

**Comment by olivermrbl** [comment]
Please add a changeset 👍

File: packages/product/src/services/product-variant.ts
Comment (adrien2p) [reply packages/product/src/services/product-variant.ts]: I would the entity, and I need to do the changes on my side as well I think

Comment (riqwan) [reply packages/product/src/services/product-variant.ts]: Meaning, TEntity right? Then we're aligned. 💪🏻

File: packages/product/integration-tests/__tests__/services/product-variant/index.ts
Comment (adrien2p) [inline packages/product/integration-tests/__tests__/services/product-variant/index.ts]: suggestion(nit): Should we store the ids when we reuse them to prevent any type and simplify any changes in the future?

File: packages/product/integration-tests/__tests__/services/product-variant/index.ts
Comment (adrien2p) [inline packages/product/integration-tests/__tests__/services/product-variant/index.ts]: todo: `should throw an error when not id is provided`?

**Review by adrien2p — APPROVED** [review]
LGTM, just a couple of things

File: .changeset/new-bottles-raise.md
Comment (adrien2p) [inline .changeset/new-bottles-raise.md]: todo: rm for now, we will add the changeset in the main pr

File: packages/product/src/services/product-variant.ts
Comment (adrien2p) [reply packages/product/src/services/product-variant.ts]: no meaning the proper entity :)

Comment (adrien2p) [reply packages/product/src/services/product-variant.ts]: if we don't touch the ouput of build query then we can use TEntity otherwise we have to use the entity, so I would say lets have only one thing instead of a mix of both

File: .changeset/new-bottles-raise.md
Comment (olivermrbl) [reply .changeset/new-bottles-raise.md]: Sorry, this is my fault. I did not realise this was branched out from a different branch than `develop`.

File: packages/product/src/services/product-variant.ts
Comment (olivermrbl) [reply packages/product/src/services/product-variant.ts]: I think it's evident that this needs documenting with very clear examples of both types of usage. @adrien2p can I get you to draft a couple of sentences about this and include in our conventions?

Comment (adrien2p) [reply packages/product/src/services/product-variant.ts]: TEntity will be used for external public API, input/output types so that any users can provide this as long as it extends the Entity. Internally, we are using the entity itself to have proper typings as we know that TEntity is at least the entity and an extends constraint is present. We can see that in the template itself where we say that the default value for TEntity is the entity itself. Does it make sense

Comment (adrien2p) [reply packages/product/src/services/product-variant.ts]: Just checked and it is already that way in the main branch btw

File: .changeset/new-bottles-raise.md
Comment (adrien2p) [reply .changeset/new-bottles-raise.md]: no problem 👍

**Pull request merged**
